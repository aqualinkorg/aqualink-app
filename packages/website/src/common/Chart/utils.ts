import type { ChartPoint } from "chart.js";
import { ChartComponentProps } from "react-chartjs-2";
import moment from "moment";
import { inRange } from "lodash";
import type { ChartProps } from ".";
import { sortByDate } from "../../helpers/sortDailyData";
import type {
  DailyData,
  SofarValue,
  SpotterData,
} from "../../store/Reefs/types";
import { SurveyListItem } from "../../store/Survey/types";

// TODO make bottom temp permanent once we work UI caveats
export const CHART_BOTTOM_TEMP_ENABLED = false;

export const filterDailyData = (
  dailyData: DailyData[],
  // Date strings, ISO preferred.
  from?: string,
  to?: string
): DailyData[] => {
  if (!from || !to) return dailyData;
  const startDate = moment(from);
  const endDate = moment(to);

  const ret = dailyData.filter((item) =>
    // add one since inRange is exclusive for last param
    inRange(moment(item.date).date(), startDate.date(), endDate.date() + 1)
  );
  // if this list is empty, it means satellite is behind. We want to display latest value, so lets just return the latest values.
  if (ret.length === 0) {
    // daily data is separated by days, so lets try match the amount of days between the range given to us.
    const diffDays = endDate.diff(startDate, "days");
    return ret.slice(-diffDays);
  }
  return ret;
};

const getSurveyDates = (surveys: SurveyListItem[]): (number | null)[] => {
  const dates = surveys.map((survey) => {
    if (survey.diveDate) {
      return new Date(survey.diveDate).setHours(0, 0, 0, 0);
    }
    return null;
  });

  return dates;
};

export const sameDay = (
  date1: string | number | Date,
  date2: string | number | Date
) => new Date(date1).toDateString() === new Date(date2).toDateString();

const timeDiff = (incomingDate: string, date: Date) =>
  Math.abs(new Date(incomingDate).getTime() - date.getTime());

export function getDailyDataClosestToDate(dailyData: DailyData[], date: Date) {
  return dailyData.reduce((prevClosest, nextPoint) =>
    timeDiff(prevClosest.date, date) > timeDiff(nextPoint.date, date)
      ? nextPoint
      : prevClosest
  );
}

export function getSpotterDataClosestToDate(
  spotterData: SofarValue[],
  date: Date,
  maxHours: number
) {
  if (spotterData.length === 0) {
    return undefined;
  }

  const closest = spotterData.reduce((prevClosest, nextPoint) =>
    timeDiff(prevClosest.timestamp, date) > timeDiff(nextPoint.timestamp, date)
      ? nextPoint
      : prevClosest
  );

  return timeDiff(closest.timestamp, date) < maxHours * 60 * 60 * 1000
    ? closest
    : undefined;
}

export const createDatasets = (
  dailyData: DailyData[],
  rawSpotterBottom: SpotterData["bottomTemperature"],
  rawSpotterSurface: SpotterData["surfaceTemperature"],
  surveys: SurveyListItem[]
) => {
  const bottomTemperature = dailyData
    .filter((item) => item.avgBottomTemperature !== null)
    .map((item) => ({
      x: item.date,
      y: item.avgBottomTemperature,
    }));

  const surfaceTemperature = dailyData
    .filter((item) => item.satelliteTemperature !== null)
    .map((item) => ({ x: item.date, y: item.satelliteTemperature }));

  const surveyDates = getSurveyDates(surveys);

  const spotterBottom = rawSpotterBottom.map((item) => ({
    x: item.timestamp,
    y: item.value,
  }));

  const spotterSurface = rawSpotterSurface.map((item) => ({
    x: item.timestamp,
    y: item.value,
  }));

  const tempWithSurvey = dailyData
    .filter(
      (item) =>
        item.satelliteTemperature !== null &&
        surveyDates.some(
          (surveyDate) => surveyDate && sameDay(surveyDate, item.date)
        )
    )
    .map((item) => ({
      x: item.date,
      y:
        // Position survey on bottom temp, if enabled, else surface temp.
        (CHART_BOTTOM_TEMP_ENABLED && item.avgBottomTemperature) ||
        item.satelliteTemperature,
    }));

  return {
    tempWithSurvey,
    bottomTemperatureData: CHART_BOTTOM_TEMP_ENABLED ? bottomTemperature : [],
    surfaceTemperatureData: surfaceTemperature,
    spotterBottom,
    spotterSurface,
  };
};

export const calculateAxisLimits = (
  dailyData: DailyData[],
  spotterBottomTemperature: SofarValue[],
  spotterSurfaceTemperature: SofarValue[],
  surveys: SurveyListItem[],
  temperatureThreshold: number | null
) => {
  const ySpacing = 1;
  const dates =
    dailyData.length > 0
      ? dailyData
          .filter(
            (item) =>
              item.surfaceTemperature !== null ||
              item.satelliteTemperature !== null
          )
          .map((item) => item.date)
      : spotterBottomTemperature.map((item) => item.timestamp);

  const spotterTimestamps = spotterBottomTemperature.map(
    (item) => item.timestamp
  );
  const spotterXMax = spotterTimestamps.slice(-1)[0];
  const spotterXMin = spotterTimestamps[0];

  const xAxisMax = spotterXMax || dates.slice(-1)[0];
  const xAxisMin = spotterXMin || dates[0];

  const {
    surfaceTemperatureData,
    bottomTemperatureData,
    spotterBottom,
    spotterSurface,
  } = createDatasets(
    dailyData,
    spotterBottomTemperature,
    spotterSurfaceTemperature,
    surveys
  );

  const temperatureData = [
    ...surfaceTemperatureData,
    ...bottomTemperatureData,
    ...spotterBottom,
    ...spotterSurface,
  ]
    .filter((value) => value)
    .map((value) => value.y);

  const yAxisMinTemp = Math.min(...temperatureData) - ySpacing;

  const yAxisMaxTemp = Math.max(...temperatureData) + ySpacing;

  const yAxisMin = Math.round(
    temperatureThreshold
      ? Math.min(yAxisMinTemp, temperatureThreshold - ySpacing)
      : yAxisMinTemp
  );

  const yAxisMax = Math.round(
    temperatureThreshold
      ? Math.max(yAxisMaxTemp, temperatureThreshold + ySpacing)
      : yAxisMaxTemp
  );

  return {
    xAxisMax,
    xAxisMin,
    yAxisMin,
    yAxisMax,
  };
};

export function useProcessedChartData(
  dailyData: ChartProps["dailyData"],
  spotterData: ChartProps["spotterData"],
  surveys: SurveyListItem[],
  temperatureThreshold: ChartProps["temperatureThreshold"],
  startDate: ChartProps["startDate"],
  endDate: ChartProps["endDate"]
) {
  // Sort daily data by date & in given date range, or latest data if no data found in range.
  const sortedFilteredDailyData = filterDailyData(
    sortByDate(dailyData, "date"),
    startDate,
    endDate
  );

  const { bottomTemperature, surfaceTemperature } = spotterData || {};

  const datasets = createDatasets(
    sortedFilteredDailyData,
    bottomTemperature || [],
    surfaceTemperature || [],
    surveys
  );

  const axisLimits = calculateAxisLimits(
    sortedFilteredDailyData,
    bottomTemperature || [],
    surfaceTemperature || [],
    surveys,
    temperatureThreshold
  );
  return { sortedFilteredDailyData, ...axisLimits, ...datasets };
}

export const createChartData = (
  spotterBottom: ChartPoint[],
  spotterSurface: ChartPoint[],
  tempWithSurvey: ChartPoint[],
  surfaceTemps: ChartPoint[],
  bottomTemps: ChartPoint[],
  fill: boolean
) => {
  const displaySpotterData = spotterSurface.length > 0;
  const data: ChartComponentProps["data"] = {
    datasets: [
      {
        type: "scatter",
        label: "SURVEYS",
        data: tempWithSurvey,
        pointRadius: 5,
        backgroundColor: "#ffffff",
        pointBackgroundColor: "#ffff",
        borderWidth: 1.5,
        borderColor: "#128cc0",
      },
      {
        label: "SURFACE TEMP",
        data: surfaceTemps,
        fill: !displaySpotterData,
        borderColor: "#6bc1e1",
        borderWidth: 2,
        pointBackgroundColor: "#ffffff",
        pointBorderWidth: 1.5,
        pointRadius: 0,
        cubicInterpolationMode: "monotone",
        backgroundColor: "rgb(107,193,225,0.2)",
      },
      {
        label: "TEMP AT DEPTH",
        data:
          CHART_BOTTOM_TEMP_ENABLED && !displaySpotterData
            ? bottomTemps
            : undefined,
        borderColor: "#46a5cf",
        borderWidth: 2,
        pointBackgroundColor: "#ffffff",
        pointBorderWidth: 1.5,
        pointRadius: 0,
        cubicInterpolationMode: "monotone",
      },
      {
        label: "SPOTTER BOTTOM",
        data: spotterBottom,
        fill: false,
        borderColor: "rgba(250, 141, 0)",
        borderWidth: 2,
        pointBackgroundColor: "#ffffff",
        pointBorderWidth: 1.5,
        pointRadius: 0,
        cubicInterpolationMode: "monotone",
      },
      {
        label: "SPOTTER SURFACE",
        data: spotterSurface,
        fill: false,
        borderColor: "#46a5cf",
        borderWidth: 2,
        pointBackgroundColor: "#ffffff",
        pointBorderWidth: 1.5,
        pointRadius: 0,
        cubicInterpolationMode: "monotone",
      },
    ],
  };

  if (fill) {
    // eslint-disable-next-line fp/no-mutating-methods
    data.datasets!.splice(1, 0, {
      label: "BLEACHING THRESHOLD",
      data: surfaceTemps,
      fill,
      borderColor: "#6bc1e1",
      borderWidth: 2,
      pointBackgroundColor: "#ffffff",
      pointBorderWidth: 1.5,
      pointRadius: 0,
      cubicInterpolationMode: "monotone",
    });
  }
  return data;
};
