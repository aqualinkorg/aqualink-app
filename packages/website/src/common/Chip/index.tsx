import React, { FC } from 'react';
import { Typography, Grid, Theme, Button } from '@mui/material';
import makeStyles from '@mui/styles/makeStyles';
import createStyles from '@mui/styles/createStyles';
import { Link } from 'react-router-dom';
import { grey, green } from '@mui/material/colors';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    chip: ({ width }: { width?: number }) => ({
      backgroundColor: grey[300],
      borderRadius: 8,
      height: 24,
      width,
      display: 'flex',
    }),
    chipText: {
      fontSize: 8,
      color: grey[600],
      [theme.breakpoints.between('md', 'lg')]: {
        fontSize: 7,
      },
    },
    circle: {
      backgroundColor: green[300],
      borderRadius: '50%',
      height: 8.4,
      width: 8.4,
      marginRight: 5,
    },
    link: {
      display: 'flex',
      alignItems: 'center',
      textDecoration: 'none',
      color: 'inherit',
      '&:hover': {
        textDecoration: 'none',
        color: 'inherit',
      },
    },
    sensorImage: {
      height: 18,
      width: 18,
    },
    button: {
      padding: 0,
      height: '100%',
    },
  }),
);

interface ChipProps {
  live?: boolean;
  href?: string;
  to?: string;
  liveText?: string;
  imageText?: string | null;
  image?: string | null;
  width?: number;
  onClick?: () => void;
  state?: any;
}

const LinkWrapper: FC<
  Pick<ChipProps, 'to' | 'href' | 'state'> & {
    className?: string;
    children?: React.ReactNode;
  }
> = ({ to, href, className, children, state }) => {
  const url = to || href;

  return url ? (
    <Link
      to={url}
      target={href ? '_blank' : undefined}
      className={className}
      rel="noopener noreferrer"
      state={state}
    >
      {children}
    </Link>
  ) : (
    <>{children}</>
  );
};

const Chip = ({
  live,
  href,
  to,
  image,
  imageText,
  liveText = 'LIVE',
  width,
  onClick,
  state,
}: ChipProps) => {
  const classes = useStyles({ width });
  return (
    <Grid className={classes.chip} item>
      <Grid container alignItems="center" justifyContent="center">
        <Button className={classes.button} onClick={onClick}>
          <LinkWrapper
            to={to}
            href={href}
            className={classes.link}
            state={state}
          >
            {live ? (
              <>
                <div className={classes.circle} />
                <Typography className={classes.chipText}>{liveText}</Typography>
              </>
            ) : (
              <>
                <Typography className={classes.chipText}>
                  {imageText}
                </Typography>
                {image && (
                  <img
                    className={classes.sensorImage}
                    alt="sensor-type"
                    src={image}
                  />
                )}
              </>
            )}
          </LinkWrapper>
        </Button>
      </Grid>
    </Grid>
  );
};

export default Chip;
