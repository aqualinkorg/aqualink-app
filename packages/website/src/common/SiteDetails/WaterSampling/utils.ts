import { formatNumber } from 'helpers/numberUtils';
import { Metrics, MetricsKeys, Sources } from 'store/Sites/types';
import { GridProps } from '@mui/material';
import siteServices from 'services/siteServices';
import { timeSeriesRequest } from 'store/Sites/helpers';
import { getSondeConfig } from 'constants/chartConfigs/sondeConfig';

type HUICardMetrics = Extract<
  Metrics,
  'salinity' | 'nitratePlusNitrite' | 'ph' | 'turbidity'
>;

export const watchColor = '#e5bb2bd0';
export const warningColor = '#ef883cd0';
export const alertColor = '#dd143ed0';

const thresholds = {
  nitratePlusNitrite: {
    good: 3.5,
    watch: 30,
    warning: 100,
  },
  turbidity: {
    good: 1,
    watch: 5,
    warning: 10,
  },
};

function getAlertColor(metric: HUICardMetrics, value?: number) {
  if (!value) return undefined;

  const compare = (th: { good: number; watch: number; warning: number }) => {
    if (value < th.good) return undefined;
    if (value < th.watch) return watchColor;
    if (value < th.warning) return warningColor;
    return alertColor;
  };

  switch (metric) {
    case 'nitratePlusNitrite':
      return compare(thresholds.nitratePlusNitrite);
    case 'turbidity':
      return compare(thresholds.turbidity);
    default:
      return undefined;
  }
}

function calculateGeometricMean(data: number[]): number | undefined {
  if (data.length === 0) return undefined;
  const lnSum = data.reduce((acc, curr) => acc + Math.log(curr), 0);
  return Math.exp(lnSum / data.length);
}

function calculateMean(data: number[]): number | undefined {
  if (data.length === 0) return undefined;
  const sum = data.reduce((acc, curr) => acc + curr);
  return sum / data.length;
}

export function getMeanCalculationFunction(
  source: Extract<Sources, 'hui' | 'sonde'>,
): (a: number[]) => number | undefined {
  switch (source) {
    case 'hui':
      return calculateGeometricMean;
    case 'sonde':
      return calculateMean;
    default:
      throw new Error(`Unknown source: ${source}`);
  }
}

const metricsForSource: Pick<
  { [Key in Sources]: MetricsKeys[] },
  'hui' | 'sonde'
> = {
  hui: ['turbidity', 'nitrate_plus_nitrite', 'ph', 'salinity'],
  sonde: [
    'odo_concentration',
    'cholorophyll_concentration',
    'ph',
    'salinity',
    'turbidity',
  ],
};

interface MetricField {
  label: string;
  value: string;
  unit?: string;
  color?: string;
  xs: GridProps['xs'];
}

export function metricFields(
  source: Extract<Sources, 'hui' | 'sonde'>,
  data?: Partial<Record<Metrics, number>>,
): MetricField[] {
  switch (source) {
    case 'hui':
      return [
        {
          label: 'Turbidity',
          value: `${formatNumber(data?.turbidity, 1)}`,
          unit: 'NTU',
          color: getAlertColor('turbidity', data?.turbidity),
          xs: 6,
        },
        {
          label: 'Nitrate Nitrite Nitrogen',
          value: `${formatNumber(data?.nitratePlusNitrite, 1)}`,
          unit: 'Âµg/L',
          color: getAlertColor('nitratePlusNitrite', data?.nitratePlusNitrite),
          xs: 6,
        },
        {
          label: 'pH',
          value: `${formatNumber(data?.ph, 1)}`,
          xs: 6,
        },
        {
          label: 'Salinity',
          value: `${formatNumber(data?.salinity, 1)}`,
          unit: source === 'hui' ? 'ppt' : 'psu',
          xs: 6,
        },
      ];

    case 'sonde':
      return [
        {
          label: 'DISSOLVED OXYGEN CONCENTRATION',
          value: formatNumber(data?.odoConcentration, 2),
          unit: getSondeConfig('odo_concentration').units,
          xs: 6,
        },
        {
          label: 'CHLOROPHYLL CONCENTRATION',
          value: formatNumber(data?.cholorophyllConcentration, 2),
          unit: getSondeConfig('cholorophyll_concentration').units,
          xs: 6,
        },
        {
          label: 'ACIDITY',
          value: formatNumber(data?.ph, 1),
          unit: getSondeConfig('ph').units,
          xs: 4,
        },
        {
          label: 'SALINITY',
          value: formatNumber(data?.salinity, 1),
          unit: getSondeConfig('salinity').units,
          xs: 5,
        },
        {
          label: 'TURBIDITY',
          value: formatNumber(data?.turbidity, 0),
          unit: getSondeConfig('turbidity').units,
          xs: 3,
        },
      ];
    default:
      throw new Error(`Unknown source: ${source}`);
  }
}

export async function getCardData(
  siteId: string,
  source: Extract<Sources, 'hui' | 'sonde'>,
) {
  try {
    const { data: uploadHistory } = await siteServices.getSiteUploadHistory(
      parseInt(siteId, 10),
    );

    const uploads =
      uploadHistory.filter(
        (x) =>
          x.dataUpload.sensorTypes.includes(source) ||
          // hui is specific type of sonde, look for hui as well when looking for sonde
          (source === 'sonde' && x.dataUpload.sensorTypes.includes('hui')),
      ) || [];
    if (uploads.length < 1) {
      return {};
    }

    switch (source) {
      case 'hui': {
        const now = new Date();
        const lastYear = now.setFullYear(now.getFullYear() - 1);
        const inLastYear = uploads.filter(
          ({ dataUpload: { maxDate } }) =>
            new Date(maxDate) > new Date(lastYear),
        );

        const minDate = inLastYear.reduce((min, curr) => {
          const currMin = curr.minDate || curr.dataUpload.minDate;
          return currMin < min ? currMin : min;
        }, new Date().toISOString());

        const maxDate =
          inLastYear.length > 0
            ? inLastYear.reduce((max, curr) => {
                const currMax = curr.maxDate || curr.dataUpload.maxDate;
                return currMax > max ? currMax : max;
              }, new Date(0).toISOString())
            : new Date().toISOString();

        const [data] = await timeSeriesRequest({
          siteId,
          start: minDate,
          end: maxDate,
          metrics: metricsForSource[source],
          hourly: true,
        });

        const pointId = inLastYear[0]?.surveyPoint;
        const samePoint =
          pointId !== null
            ? inLastYear.reduce(
                (acc, curr) => acc && curr.surveyPoint?.id === pointId.id,
                true,
              )
            : false;

        return {
          data,
          minDate,
          maxDate,
          point: samePoint ? pointId : undefined,
        };
      }
      case 'sonde': {
        // eslint-disable-next-line fp/no-mutating-methods
        uploads.sort((a, b) => {
          if (a.maxDate && b.maxDate) {
            const timeA = new Date(a.maxDate).getTime();
            const timeB = new Date(b.maxDate).getTime();
            return timeB - timeA;
          }
          return 0;
        });
        const { minDate, maxDate } = uploads[0].dataUpload;
        const { surveyPoint } = uploads[0];
        const [data] = await timeSeriesRequest({
          siteId,
          start: minDate,
          end: maxDate,
          metrics: metricsForSource[source],
          hourly: true,
        });
        return {
          data,
          minDate,
          maxDate,
          point: surveyPoint,
        };
      }
      default:
        throw new Error(`Unknown source: ${source}`);
    }
  } catch (err) {
    console.error(err);
    return {};
  }
}
