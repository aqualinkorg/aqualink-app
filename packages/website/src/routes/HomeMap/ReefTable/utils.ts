import { startCase } from "lodash";
import type { TableRow } from "../../../store/Homepage/types";

export type Order = "asc" | "desc";

export enum OrderKeys {
  ALERT = "alert",
  DEPTH = "depth",
  DHW = "dhw",
  LOCATION_NAME = "locationName",
  TEMP = "temp",
}

// This type isn't used anywhere, it just forces the above enum to only hold valid keys to TableRow.
type __CheckOrderKeys = TableRow[OrderKeys];

/**
 * Returns a human friendly string of a order key.
 * @param orderKey key to return the string for.
 */
export function getOrderKeysFriendlyString(orderKey: OrderKeys) {
  switch (orderKey) {
    case OrderKeys.DHW:
      return "DHW";
    default:
      return startCase(orderKey);
  }
}

export function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  return b[orderBy] >= a[orderBy] ? 1 : -1;
}

export function getComparator(
  order: Order,
  orderBy: OrderKeys
): (
  a: {
    [key in OrderKeys]: number | string | null;
  },
  b: {
    [key in OrderKeys]: number | string | null;
  }
) => number {
  return order === "desc"
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy);
}

export function stableSort<T>(array: T[], comparator: (a: T, b: T) => number) {
  // eslint-disable-next-line fp/no-mutating-methods
  return [...array].sort(comparator);
}
