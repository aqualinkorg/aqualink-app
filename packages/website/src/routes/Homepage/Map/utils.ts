export const coloredBuoy = (color: string) =>
  `<?xml version="1.0" encoding="UTF-8"?>
  <svg width="29px" height="35px" viewBox="0 0 29 35" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
      <defs>
          <path d="M10.0023993,-6.25277607e-13 C17.4094117,-6.25277607e-13 22.207967,7.82436546 18.9739831,14.3966715 C17.4094871,17.7347423 12.1924976,22.7424773 10.0023993,25.1417547 C7.70723368,22.5332976 2.80456638,17.9439217 1.13490246,14.5007584 C-2.30725517,8.03253919 2.49117441,-6.25277607e-13 10.0023993,-6.25277607e-13 Z" id="path-1"></path>
          <filter x="-35.0%" y="-19.9%" width="170.0%" height="155.7%" filterUnits="objectBoundingBox" id="filter-2">
              <feOffset dx="0" dy="2" in="SourceAlpha" result="shadowOffsetOuter1"></feOffset>
              <feGaussianBlur stdDeviation="2" in="shadowOffsetOuter1" result="shadowBlurOuter1"></feGaussianBlur>
              <feColorMatrix values="0 0 0 0 0   0 0 0 0 0   0 0 0 0 0  0 0 0 0.5 0" type="matrix" in="shadowBlurOuter1"></feColorMatrix>
          </filter>
      </defs>
      <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
          <g id="Reef-Detail-full---no-live-feed-Copy-2" transform="translate(-748.000000, -491.000000)">
              <g id="icon_sat" transform="translate(752.500000, 493.500000)">
                  <g id="Fill-1">
                      <use fill="black" fill-opacity="1" filter="url(#filter-2)" xlink:href="#path-1"></use>
                      <use fill="${color}" fill-rule="evenodd" xlink:href="#path-1"></use>
                  </g>
                  <path d="M15.2680598,9.87791599 C11.5105654,13.6354103 6.18310234,8.31883808 9.9461188,4.555975 L12.1649364,6.77540616 L12.1649364,6.41409093 C12.8785627,6.41409093 13.4591827,6.99710385 13.4591827,7.7107302 L13.0996622,7.70953221 L15.2680598,9.87791599 Z M16.5730897,7.82384158 C16.5736894,5.33842841 14.5375756,3.3 12.0515489,3.3 L12.0515489,4.19818779 C14.0402783,4.19818779 15.6772794,5.83150751 15.6772794,7.82085043 L16.5730897,7.82324642 L16.5730897,7.82384158 Z M15.0787125,7.78729129 C15.0787125,6.1436943 13.7305335,4.79553061 12.0869518,4.79553061 L12.0869518,5.69431663 C13.2361966,5.69431663 14.1799878,6.63803121 14.1799878,7.78735265 L15.0787125,7.78729129 Z M14.2254993,14.820015 L13.1571229,15.8883607 C12.6082704,16.4372131 11.7100826,16.4372131 11.1611841,15.8883607 L9.19592376,13.9231004 L11.2625763,11.8564478 L14.2254993,14.8193708 L14.2254993,14.820015 Z M4.90689491,5.50141065 L3.8385495,6.56978704 C3.28969701,7.11863953 3.28969701,8.01682732 3.8385495,8.56572583 L5.80380954,10.5309862 L7.87046211,8.46433361 L4.90753915,5.50141065 L4.90689491,5.50141065 Z M7.7537,13.1817868 L6.98793566,13.9475511 C6.76803266,14.1668559 6.40912568,14.1668559 6.18922268,13.9475511 L5.77877594,13.5371044 C5.55947117,13.3172014 5.55947117,12.9582944 5.77877594,12.7383914 L6.54454028,11.9726271 C6.58228856,12.0427319 6.63082337,12.1086429 6.68954343,12.1673584 L7.55837196,13.0361869 C7.61709201,13.094907 7.68300309,13.1434418 7.75250507,13.1811901 L7.7537,13.1817868 Z M10.5345618,11.2584025 L9.13185186,12.6610971 C8.80230414,12.9906448 8.26362156,12.9906448 7.93407384,12.6610971 L7.06524531,11.7922685 C6.73569759,11.4627208 6.73569759,10.9240382 7.06524531,10.5944905 L8.50748441,9.15225143 C8.86340024,10.0426469 9.65732738,10.8653351 10.5345618,11.2578656 L10.5345618,11.2584025 Z" id="Fill-1" fill="#FFFFFF" transform="translate(10.000000, 9.800000) scale(-1, -1) translate(-10.000000, -9.800000) "></path>
              </g>
          </g>
      </g>
  </svg>`;
