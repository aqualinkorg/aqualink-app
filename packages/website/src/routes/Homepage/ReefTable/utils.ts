export type Order = "asc" | "desc";

export type OrderKeys = "locationName" | "temp" | "depth" | "dhw" | "alert";

export function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  return b[orderBy] >= a[orderBy] ? 1 : -1;
}

export function getComparator(
  order: Order,
  orderBy: OrderKeys
): (
  a: {
    [key in OrderKeys]: number | string | null;
  },
  b: {
    [key in OrderKeys]: number | string | null;
  }
) => number {
  return order === "desc"
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy);
}

export function stableSort<T>(array: T[], comparator: (a: T, b: T) => number) {
  return [...array].sort(comparator);
}
