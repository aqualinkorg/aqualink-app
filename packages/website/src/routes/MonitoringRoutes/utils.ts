import { OptionsObject, SnackbarKey, SnackbarMessage } from 'notistack';
import React from 'react';
import { User } from 'store/User/types';

interface FetchDataProps<T> {
  user: User | null;
  enqueueSnackbar: (
    message: SnackbarMessage,
    options?: OptionsObject | undefined,
  ) => SnackbarKey;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setResult?: React.Dispatch<React.SetStateAction<T | null>>;
  getResult: (token: string) => Promise<T>;
}

export async function fetchData<T>({
  user,
  enqueueSnackbar,
  setLoading,
  setResult,
  getResult,
}: FetchDataProps<T>) {
  const { token } = user || {};
  if (!token) {
    enqueueSnackbar('User is not authenticated!', {
      variant: 'error',
    });
    return;
  }

  if (setResult) {
    setResult(null);
  }

  setLoading(true);
  try {
    const data = await getResult(token);

    if (setResult) {
      setResult(data);
    }
  } catch (error: any) {
    const message = error?.response?.data?.message || error?.message;
    const errorMessage =
      typeof message === 'object' ? message?.join('. ') : message;
    enqueueSnackbar(errorMessage || 'Request failed', {
      variant: 'error',
    });
  } finally {
    setLoading(false);
  }
}

export const includes = (
  a: string | undefined | null,
  b: string | undefined,
) => {
  if (!b?.trim()) return true;
  if (!a) return false;

  return a.toLocaleLowerCase().includes(b.toLocaleLowerCase());
};
