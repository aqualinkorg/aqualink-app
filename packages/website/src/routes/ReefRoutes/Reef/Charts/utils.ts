import type { Data } from "../../../../store/Reefs/types";

export const createDatasets = (dailyData: Data[]) => {
  const dailyDataLen = dailyData.length;

  const bottomTemperature = dailyData.map((item) => item.avgBottomTemperature);
  const surfaceTemperature = dailyData
    .filter(
      (item) =>
        item.surfaceTemperature !== null || item.satelliteTemperature !== null
    )
    .map((item) => item.surfaceTemperature || item.satelliteTemperature);

  const meanBottomTemperature =
    bottomTemperature.reduce((a, b) => a + b) / dailyDataLen;
  const meanSurfaceTemperature =
    surfaceTemperature.reduce((a, b) => a + b) / dailyDataLen;

  return {
    bottomTemperatureData: [
      meanBottomTemperature,
      ...bottomTemperature,
      meanBottomTemperature,
    ],
    surfaceTemperatureData: [
      meanSurfaceTemperature,
      ...surfaceTemperature,
      meanSurfaceTemperature,
    ],
  };
};

export const calculateAxisLimits = (dailyData: Data[]) => {
  const dates = dailyData
    .filter(
      (item) =>
        item.surfaceTemperature !== null || item.satelliteTemperature !== null
    )
    .map((item) => item.date);
  const dailyDataLen = dates.length;

  const xAxisMax = new Date(
    new Date(dates[dailyDataLen - 1]).setHours(24, 0, 0, 0)
  ).toISOString();

  const xAxisMin = new Date(
    new Date(dates[0]).setHours(-1, 0, 0, 0)
  ).toISOString();

  // Add an extra date one day after the final daily data date
  const chartLabels = [
    new Date(new Date(xAxisMin).setHours(3, 0, 0, 0)).toISOString(),
    ...dates,
    new Date(new Date(xAxisMax).setHours(3, 0, 0, 0)).toISOString(),
  ];

  return {
    xAxisMax,
    xAxisMin,
    chartLabels,
  };
};
