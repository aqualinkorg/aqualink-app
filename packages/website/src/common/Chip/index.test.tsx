import React from 'react';
import { renderWithProviders } from 'utils/test-utils';
import Chip from '.';

test('renders as expected', () => {
  const { container } = renderWithProviders(<Chip live />);
  expect(container).toMatchSnapshot();
});
