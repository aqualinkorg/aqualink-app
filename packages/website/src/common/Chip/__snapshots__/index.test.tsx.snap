// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item makeStyles-chip-1 makeStyles-chip-7 css-x7bhrq-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-6zo1m7-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary makeStyles-button-6 css-iorydh-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <div
          class="makeStyles-circle-3"
        />
        <mock-typography
          classname="makeStyles-chipText-2"
        >
          LIVE
        </mock-typography>
      </button>
    </div>
  </div>
</div>
`;
